File: insertion_sort.py
# filename: insertion_sort.py
import random
rng = 20 # default value
rng = int(input("Please input a value of range: "))
lst = [random.randrange(0, rng) for idx in range(rng)]

# insertion sort
sort_lst = lst[:1]
for idx in range(1, rng):
    for i in range(idx):
        if lst[idx] <= sort_lst[i]:
            sort_lst.insert(i, lst[idx])
            break
    else:
        sort_lst.append(lst[idx])

# print the original-list and sorted-list
print("=" * 40)
print("The original-list:\n", lst)
print("The sorted-list:\n", sort_lst)
lst.sort()
print("Compare with result sorted by list.sort():\n", lst)
print("Is the sort-algorithm right?\n", lst == sort_lst)
print("--> THIS IS INSERTION SORT! <--")

File: clang-algorithm/sort_algorithm/insertion_sort/insertion_sort.c
// insertion sort
#include <stdio.h>
int main(void){
    int data[6] = {77, 19, 80, 79, 20, 11};
    int i, j, key;
    int len = 6;

    for (i = 1; i < len; i++){
        // 保存待插入数据
        key = data[i];
        // 在有序序列中从右往左找插入位置
        j = i - 1;
        // 从右往左的好处是每次遇到比待插数据大的数时往右移动
        while (j >= 0 && data[j] > key){
            data[j + 1] = data[j];
            j--;
        }
        // 找到插入位置后放入带插数据
        data[j + 1] = key;
    }

    for (i = 0; i < len; i++){
        printf("%d ", data[i]);
    }
    putchar('\n');

    return 0;
}
